crates/dojo-core/src/database.cairo

```diff
@@ -38,36 +38,87 @@ fn set(
 }
 
 fn set_with_index( 
-    class_hash: starknet::ClassHash, table: felt252, key: felt252, offset: u8, value: Span<felt252>, layout: Span<u8>
+    class_hash: starknet::ClassHash,
+    table: felt252,
+    key: felt252,
+    keys: Span<felt252>,
+    offset: u8,
+    value: Span<felt252>,
+    layout: Span<u8>
 ) {
     set(class_hash, table, key, offset, value, layout);
-    index::create(0, table, key);
+    index::create(0, table, key, 0); // create a record in index of all records
+
+    let mut idx = 0;
+    loop {
+        if idx == keys.len() {
+            break;
+        }
+        let mut serialized = ArrayTrait::new();
+        table.serialize(ref serialized);
+        idx.serialize(ref serialized);
+        let index = poseidon_hash_span(serialized.span());
+
+        index::create(0, index, key, *keys.at(idx)); // create a record for each of the keys
+        
+        idx += 1;
+    };
+
+    let len_keys = array!['dojo_storage_keys_len', table, key].span();
+    storage::set(0, len_keys, keys.len().into()); // save the number of keys
 }
 
 fn del(class_hash: starknet::ClassHash, table: felt252, key: felt252) {
     index::delete(0, table, key);
+
+    let len_keys = array!['dojo_storage_keys_len', table, key].span();
+    let len = storage::get(0, len_keys);
+
+    let mut idx = 0;
+    loop {
+        if idx == len {
+            break;
+        }
+        let mut serialized = ArrayTrait::new();
+        table.serialize(ref serialized);
+        idx.serialize(ref serialized);
+        let index = poseidon_hash_span(serialized.span());
+
+        index::delete(0, index, key);
+        
+        idx += 1;
+    };
+
+    storage::set(0, len_keys, 0); // overwrite the number of keys
 }
 
 // Query all entities that meet a criteria. If no index is defined,
 // Returns a tuple of spans, first contains the entity IDs,
 // second the deserialized entities themselves.
 fn scan(
-    class_hash: starknet::ClassHash, model: felt252, where: Option<WhereCondition>, values_length: usize, values_layout: Span<u8>
+    class_hash: starknet::ClassHash, model: felt252, index: Option<felt252>, where: Option<WhereCondition>, values_length: usize, values_layout: Span<u8>
 ) -> (Span<felt252>, Span<Span<felt252>>) {
     match where {
         Option::Some(clause) => {
             let mut serialized = ArrayTrait::new();
             model.serialize(ref serialized);
             clause.key.serialize(ref serialized);
-            let index = poseidon_hash_span(serialized.span());
+            let table = poseidon_hash_span(serialized.span());
+
+            let ids = match index {
+                Option::Some(index) => match index::get_at(0, table, clause.value, index) {
+                    Option::Some(id) => array![id],
```

Comment:
Hm why does this return only a single entry?

Reply:
/// * `index` - The index to be retrieved.
I interpreted this comment in proposed API as a single position in the index that is to be returned.
Could you specify what was the intention? Should index be an alternative to `query_key`?